File: src/ci/py/createBadges.py
import anybadge
import os
import json

covBadgeSvg = 'cov-badge.svg'
testBadgeSvg = 'test-badge.svg'

with open('coverage.json') as jsonFile:
    covJson = json.load(jsonFile)
    print(covJson["totals"]["percent_covered"])

covPct = covJson["totals"]["percent_covered"]
if covPct < 20:
    color = 'red'
elif covPct < 70:
    color = 'orange_2'
else:
    color = 'green'

badge = anybadge.Badge('coverage', str(int(round(covPct))) + '%', default_color=color)

if os.path.exists(covBadgeSvg):
    os.remove(covBadgeSvg)
badge.write_badge(covBadgeSvg)

with open('test-report.json') as jsonFile:
    testJson = json.load(jsonFile)

testSummary = testJson["report"]["summary"]

testPassed = 0 if "passed" not in testSummary else testSummary["passed"]
testFailed = 0 if "failed" not in testSummary else testSummary["failed"]
print(str(testPassed) + " " + str(testFailed))

color = 'green' if (testPassed != 0 and testFailed == 0) else 'red'

badge = anybadge.Badge('tests', str(testPassed) + " passed, " + str(testFailed) + " failed", default_color=color)
if os.path.exists(testBadgeSvg):
    os.remove(testBadgeSvg)
badge.write_badge(testBadgeSvg)

File: src/ci/js/createBadges.mjs
// Create badge icons for tests and coverage.
// Usage: node createBadges.mjs <coverage-badge-filename> <tests-badge-filename>

import fs  from 'fs'
import { makeBadge } from 'badge-maker'

import { createRequire } from 'module'
const require = createRequire(import.meta.url)

// Create coverage badge
const covSummary = require('../coverage/coverage-summary.json')
const covPct = covSummary.total.statements.pct

let color = 'green'
if (covPct < 20) color = 'red'
else if (covPct < 70) color = '#ffa500' // orange_2

const covBadge = {
    label: 'coverage',
    message: covPct.toFixed(0) + '%',
    color: color
}
const covSvg = makeBadge(covBadge)

// Create badge for shields.io
// const covBadge = {
//     schemaVersion: 1,
//     label: 'coverage',
//     message: covPct.toFixed(0) + '%',
//     color: color
// }
// console.log("covBadge:", JSON.stringify(covBadge))
// fs.writeFileSync(process.argv[2], JSON.stringify(covBadge))

// Create test results badge
const testSummary = require('../test-results.json')
const numPassedTests = testSummary.numPassedTests
const numFailedTests = testSummary.numFailedTests
color = numPassedTests != 0 && numFailedTests == 0? 'green' : 'red'

const testBadge = {
    label: 'tests',
    message: numPassedTests + " passed, " + numFailedTests + " failed",
    color: color
}
const testSvg = makeBadge(testBadge)

// Write badge files.
let covBadgeFile = process.argv[2] == undefined? "cov-badge.svg" : process.argv[2]
let testBadgeFile = process.argv[3] == undefined? "test-badge.svg" : process.argv[3]
fs.writeFile(covBadgeFile, covSvg, err => {
    if (err) {
        console.log(`Error writing ${covBadgeFile}: ${err.message}`)
        process.exit(1)
    }
    fs.writeFile(testBadgeFile, testSvg, err => {
        if (err) {
            console.log(`Error writing ${testBadgeFile}: ${err.message}`)
            process.exit(1)
        }
        console.log(`Created ${covBadgeFile} and ${testBadgeFile}`)
    })
})
